File: backend/google_places.py
from fastapi import Body
import os
import requests
from fastapi import APIRouter, Query
from fastapi.responses import JSONResponse, RedirectResponse

GOOGLE_API_KEY = os.getenv("GOOGLE_PLACES_API_KEY")
router = APIRouter()

@router.get("/api/places/nearby")
def nearby_search(lat: float = Query(...), lng: float = Query(...), radius: int = Query(1000)):
    url = "https://maps.googleapis.com/maps/api/place/nearbysearch/json"
    params = {
        "location": f"{lat},{lng}",
        "radius": radius,
        "type": "restaurant",
        "key": GOOGLE_API_KEY,
    }
    resp = requests.get(url, params=params)
    data = resp.json()
    # Return only basic info and place IDs for frontend
    restaurants = []
    for r in data.get("results", []):
        restaurants.append({
            "name": r.get("name"),
            "lat": r.get("geometry", {}).get("location", {}).get("lat"),
            "lng": r.get("geometry", {}).get("location", {}).get("lng"),
            "place_id": r.get("place_id"),
            "rating": r.get("rating"),
            "price_level": r.get("price_level"),
            "vicinity": r.get("vicinity"),
            "types": r.get("types", []),
            "photo_reference": r.get("photos", [{}])[0].get("photo_reference") if r.get("photos") else None
        })
    return JSONResponse(content={"restaurants": restaurants})

@router.post("/api/places/details/batch")
def batch_place_details(place_ids: list = Body(...)):
    url = "https://maps.googleapis.com/maps/api/place/details/json"
    details = []
    for pid in place_ids:
        params = {
            "place_id": pid,
            "key": GOOGLE_API_KEY,
        }
        resp = requests.get(url, params=params)
        data = resp.json()
        if data.get("status") == "OK":
            details.append(data.get("result"))
    return JSONResponse(content={"details": details})
@router.get("/api/places/details")
def place_details(place_id: str = Query(...)):
    url = "https://maps.googleapis.com/maps/api/place/details/json"
    # Request specific fields for detailed restaurant information
    fields = [
        "place_id", "name", "formatted_address", "formatted_phone_number",
        "website", "rating", "price_level", "opening_hours", "reviews",
        "photos", "geometry", "types", "vicinity", "business_status"
    ]
    params = {
        "place_id": place_id,
        "fields": ",".join(fields),
        "key": GOOGLE_API_KEY,
    }
    resp = requests.get(url, params=params)
    data = resp.json()
    
    # Format the response to match our expected structure
    if data.get("status") == "OK" and "result" in data:
        result = data["result"]
        formatted_result = {
            "place_id": result.get("place_id"),
            "name": result.get("name"),
            "formatted_address": result.get("formatted_address"),
            "formatted_phone_number": result.get("formatted_phone_number"),
            "website": result.get("website"),
            "rating": result.get("rating"),
            "price_level": result.get("price_level"),
            "vicinity": result.get("vicinity"),
            "business_status": result.get("business_status"),
            "types": result.get("types", []),
            "geometry": result.get("geometry"),
            "opening_hours": result.get("opening_hours"),
            "reviews": result.get("reviews", [])[:5],  # Limit to first 5 reviews
            "photos": result.get("photos", [])[:3],  # Limit to first 3 photos
        }
        return JSONResponse(content={"result": formatted_result, "status": "OK"})
    
    return JSONResponse(content=data)

@router.get("/api/places/photo")
def place_photo(photo_reference: str = Query(...), maxwidth: int = Query(400)):
    if not GOOGLE_API_KEY:
        return JSONResponse(content={"error": "Google API key not configured"}, status_code=500)

    url = "https://maps.googleapis.com/maps/api/place/photo"
    params = {
        "photoreference": photo_reference,
        "maxwidth": maxwidth,
        "key": GOOGLE_API_KEY,
    }

    try:
        resp = requests.get(url, params=params, allow_redirects=False)

        if resp.status_code == 302 and "Location" in resp.headers:
            return RedirectResponse(resp.headers["Location"])
        elif resp.status_code == 200:
            return RedirectResponse(url=resp.url)
        else:
            return JSONResponse(content={"error": f"Photo not found, status: {resp.status_code}, body: {resp.text}"}, status_code=404)
    except Exception as e:
        return JSONResponse(content={"error": "Failed to fetch photo"}, status_code=500)

File: lib/google-places.ts
// Google Places API utilities

export interface Restaurant {
  place_id: string
  name: string
  geometry: {
    location: {
      lat: number
      lng: number
    }
  }
  vicinity: string
  photo_reference?: string
  rating?: number
  price_level?: number
  types?: string[]
}

export interface RestaurantDetails {
  place_id: string
  name: string
  formatted_address?: string
  formatted_phone_number?: string
  website?: string
  rating?: number
  price_level?: number
  vicinity: string
  business_status?: string
  types?: string[]
  geometry: {
    location: {
      lat: number
      lng: number
    }
  }
  opening_hours?: {
    open_now?: boolean
    periods?: Array<{
      close?: { day: number; time: string }
      open?: { day: number; time: string }
    }>
    weekday_text?: string[]
  }
  reviews?: Array<{
    author_name: string
    author_url?: string
    language?: string
    profile_photo_url?: string
    rating: number
    relative_time_description: string
    text: string
    time: number
  }>
  photos?: Array<{
    height: number
    width: number
    photo_reference: string
  }>
}

export interface PlacesSearchRequest {
  location: { lat: number; lng: number }
  radius: number
  type: string
}

export class GooglePlacesService {
  private apiKey: string
  private static cache = new Map<string, { data: Restaurant[], timestamp: number }>()
  private static CACHE_DURATION = 5 * 60 * 1000 // 5 minutes in milliseconds

  constructor(apiKey: string) {
    this.apiKey = apiKey
  }

  private getCacheKey(request: PlacesSearchRequest): string {
    return `${request.location.lat.toFixed(4)},${request.location.lng.toFixed(4)},${request.radius}`
  }

  private isValidCache(timestamp: number): boolean {
    return Date.now() - timestamp < GooglePlacesService.CACHE_DURATION
  }

  async searchNearbyRestaurants(request: PlacesSearchRequest): Promise<Restaurant[]> {
    try {
      if (!request.location || typeof request.location.lat !== "number" || typeof request.location.lng !== "number") {
        console.warn("searchNearbyRestaurants called without valid location", request.location)
        return []
      }

      // Check cache first
      const cacheKey = this.getCacheKey(request)
      const cachedEntry = GooglePlacesService.cache.get(cacheKey)
      
      if (cachedEntry && this.isValidCache(cachedEntry.timestamp)) {
        console.log('🎯 Service cache hit! Returning cached restaurants')
        return cachedEntry.data
      }

      console.log('🌐 Service cache miss. Fetching from backend...')
      
      // In a real implementation, this would call Google Places API
      const response = await fetch(
        `http://localhost:8000/api/places/nearby?lat=${request.location.lat}&lng=${request.location.lng}&radius=${request.radius}`
      )

      if (!response.ok) {
        throw new Error("Failed to fetch restaurants")
      }

      const data = await response.json()
      
      // Cache the result
      GooglePlacesService.cache.set(cacheKey, {
        data: data.restaurants,
        timestamp: Date.now()
      })
      
      console.log('💾 Cached restaurants for key:', cacheKey)
      
      return data.restaurants
    } catch (error) {
      console.error("Error fetching restaurants:", error)
      return []
    }
  }

  async getPlaceDetails(placeId: string): Promise<Restaurant | null> {
    try {
      // In a real implementation, this would call Google Places API Place Details
  const response = await fetch(`http://localhost:8000/api/places/details?place_id=${placeId}`)

      if (!response.ok) {
        throw new Error("Failed to fetch place details")
      }

      const data = await response.json()
      return data.restaurant
    } catch (error) {
      console.error("Error fetching place details:", error)
      return null
    }
  }
}

// Utility function to calculate distance between two points
export function calculateDistance(lat1: number, lng1: number, lat2: number, lng2: number): string {
  const R = 3959 // Earth's radius in miles
  const dLat = ((lat2 - lat1) * Math.PI) / 180
  const dLng = ((lng2 - lng1) * Math.PI) / 180
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos((lat1 * Math.PI) / 180) * Math.cos((lat2 * Math.PI) / 180) * Math.sin(dLng / 2) * Math.sin(dLng / 2)
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))
  const distance = R * c

  return distance < 1 ? `${(distance * 5280).toFixed(0)} ft` : `${distance.toFixed(1)} mi`
}
